apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,428 @@
+      } else {
+        window.dispatchEvent(new CustomEvent('bluetooth-disabled'));
+      }
+
+      this._stateChangeHandler =
+        this._bluetoothAdapterHandler.bind(this);
+      this._adapter.addEventListener('attributechanged',
+        this._stateChangeHandler);
+    }
+
+    /*
+     * In file transfering case:
+     * since System Message can't be listened in two js files within a app,
+     * so we listen here but dispatch events to bluetooth_transfer.js
+     * while getting bluetooth file transfer start/complete system messages
+     */
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-start',
+      (transferInfo) => {
+        this._setProfileConnected(this.Profiles.OPP, true);
+        window.dispatchEvent(new CustomEvent('bluetooth-opp-transfer-start',
+          {transferInfo: transferInfo}));
+      }
+    );
+
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-complete',
+      (transferInfo) => {
+        this._setProfileConnected(this.Profiles.OPP, false);
+        window.dispatchEvent(new CustomEvent('bluetooth-opp-transfer-complete',
+          {transferInfo: transferInfo}));
+      }
+    );
+
+    // decouple bluetooth enable/disable function from other system part
+    window.addEventListener('request-enable-bluetooth',
+      this._requestEnableHandler.bind(this));
+    window.addEventListener('request-disable-bluetooth',
+      this._requestDisableHandler.bind(this));
+
+    Service.registerState('isEnabled', this);
+    // LazyLoader.load(['js/bluetooth_icon.js',
+    //                  'js/bluetooth_transfer_icon.js',
+    //                  'js/bluetooth_headphone_icon.js'], function() {
+    //   this.icon = new BluetoothIcon(this);
+    //   this.icon.start();
+    //   this.transferIcon = new BluetoothTransferIcon(this);
+    //   this.transferIcon.start();
+    //   this.headphoneIcon = new BluetoothHeadphoneIcon(this);
+    //   this.headphoneIcon.start();
+    // }.bind(this)).catch(function(err) {
+    //   console.error(err);
+    // });
+  },
+
+  /**
+   * Remove all EventListeners and adapters when default adapter
+   * is changed.
+   */
+  _cleanupListenerAndAdapters() {
+    this._bluetooth.removeEventListener('attributechanged',
+      this._defaultAdapterChangeHandler);
+    this._adapter.removeEventListener('attributechanged',
+      this._stateChangeHandler);
+    // remove getAdapter cached promise
+    if (this._cacheGetAdapterPromise) {
+      this._bluetooth.removeEventListener('attributechanged',
+        this._promiseAdapterChangeHandler);
+      this._cacheGetAdapterPromise = null;
+    }
+
+    this._adapter = null;
+  },
+
+  /**
+   * update settings value and enable bluetooth hardware
+   *
+   * @private
+   */
+  _requestEnableHandler: function bt__requestEnableHandler() {
+    this._debug('enabling bluetooth');
+    this.getAdapter().then((adapter) => {
+      adapter.enable().then(() => { //resolve
+        this._debug('bluetooth enabled');
```

Comment:
If the call succeeds, `bluetooth.enabled` should also be updated because in v2 gaia is responsible for maintaining the setting.